Update docs (#40)
psychoinformatics-de/inm-icf-utilities: docs/source/user/datalad-generate.rst

DataLad datasets, which index content and location of these tarballs,
for DataLad-based access on institute-local infrastructure.

In principle, such datasets are *lightweight*, meaning that they only
index the content that can be retrieved from the ICF archive (all
access restrictions apply). Using DataLad can simplify local access,
allow raw data versioning, integrate with existing workflows, and
enable logical transformations of the DICOM folder structure - see
:ref:`dl-advanced` for examples of the latter.

The workflow described below uses DataLad with DataLad-Next extension
for initial DICOM download and the INM-ICF tools packaged as a
Singularity container for subsequent steps (see
:ref:`dl-requirements`). ICF access credentials are required (see
:ref:`dl-credentials`).

Obtain the tarball
^^^^^^^^^^^^^^^^^^

First, create an empty directory to be the local dataset store. The
last path component must be the ``project-ID`` used by the ICF store,
because following commands use project and visit IDs to determine
paths.

.. code-block:: bash

   mkdir -p local_dicom_store/<project-ID>

Download the visit tarball, keeping the same relative path:

.. code-block:: bash

>    datalad download "https://data.inm-icf.de/<project-ID>/<visit-ID>_dicom.tar <project-ID>/<visit-ID>_dicom.tar"

Reviewer: I believe it should go inside the local store?

```suggestion
   datalad download "https://data.inm-icf.de/<project-ID>/<visit-ID>_dicom.tar  local_dicom_store/<project-ID>/<visit-ID>_dicom.tar"
```

Author: Lol, the double space in the argument makes it download to ` local_dicom_store` instead of `local_dicom_store`.

Reviewer: > Lol, the double space in the argument makes it download to  local_dicom_store instead of local_dicom_store.

oh no.... :o

Using ``datalad download`` for downloading the file has the benefit of
using DataLad's credential management. If this is the first time you
use DataLad to access the project directory, you will be asked to
provide your ICF credentials. See :ref:`dl-credentials` for details.

For the following examples, the *absolute path* to the local DICOM
store will be represented by ``$STORE_DIR``:

.. code-block:: bash

   export STORE_DIR=$PWD/local_dicom_store

Deposit visit metadata alongside tarball
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Information required to create a DataLad dataset needs to be extracted
from the tarball:

> 

Reviewer: Following through the docs sequentially, I don't think I've come across this singularity image before. I think it would make sense to link to its download page here.

Author: The 3rd paragraph of this page says:

> The workflow described below uses DataLad with DataLad-Next extension for initial DICOM download and the INM-ICF tools packaged as a Singularity container for subsequent steps (see DataLad requirements).

where "DataLad requirements" is a link to a page that describes things in greater details (and is actually positioned earlier in the User Guide), and links to containerized execution page.

However, your comment makes it apparent that I didn't do a good enough job when trying to compartmentalize the docs (to avoid repetition), and I will add a sentence of two to make up for it.

<hr>

By the way, this points to a small design issue with the tooling. Initially, the Singularity image was just for ICF. ICF would only use DataLad through the scripts in this image. Users would not need the Singularity image, they would clone datasets from ICF using DataLad.

Now, users who want to dataladify datasets using the Singularity image still need to download the tarballs somehow. I decided to suggest `datalad download` for the task, because it interacts with DataLad credentials, that would also be needed for any subsequent dataset content retrieval from ICF. Alternatively, we could recommend `curl -u` followed by Singularity (no need to install DataLad), or `datalad download` followed by running scripts from this repo (no need for Singularity). The former seems unsatisfactory, because any further dataset interaction would need to happen through DataLad anyway. The latter seems unsatisfactory because the Singularity image was introduced to make the ICF tooling independent of changes in DataLad.

.. code-block:: bash

   singularity run -B $STORE_DIR icf.sif deposit_visit_metadata \
     --store-dir $STORE_DIR --id <project-ID> <visit ID>

This will generate two files, ``<visit ID>_metadata_dicoms.json`` and
``<visit ID>_metadata_tarball.json``, and place them alongside the
tarball. The former contains metadata describing individual files
within the tarball (relative path, MD5 checksum, size, and a small
subset of DICOM headers describing acquisition type), and the latter
describes the tarball itself.

Deposit dataset alongside tarball
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> DataLad dataset is created based on the metadata extracted in the

Reviewer: ?

```suggestion
A DataLad dataset is created based on the metadata extracted in the
```

Reviewer: Re-reading this paragraph many times, I feel like I'm not 100% sure what it is telling me. Maybe one introductory sentence in addition helps. Is the gist something like this?

```suggestion
In order to deposit a DataLad dataset next to the original tarball in the remote data store, the following command creates a DataLad dataset  based on the metadata extracted in the
```

Reviewer: I think the command also misses the ``--id`` parameter and placeholders? I'm getting this when running it:

```
(icf) adina@muninn in /tmp
❱ singularity run -B $STORE_DIR icf.sif deposit_visit_dataset \
  --store-dir $STORE_DIR --store-url https://data.inm-icf.de
usage: deposit_visit_dataset [-h] --id STUDY-ID VISIT-ID [-o PATH] [--store-url URL]
deposit_visit_dataset: error: the following arguments are required: --id
```

Reviewer: sorry for the flood of comments, I'm realizing more and more things as I'm walking through - I was expecting this to generate a dataset based on the heading, but it doesn't create a standard dataset on my system - just the lightweight representation. Maybe we can reflect this in the heading and description, eg with by placing "dataset" in air quotes or calling it lightweight dataset representation already at the start?

Author: No need to apologize; thanks a lot for these comments. I agree with the points you make and will make changes accordingly (without using the suggestions directly).

previous step.  Additionally, you need to provide the base URL of the
ICF store, ``<ICF STORE URL>`` (this base URL should not contain study
or visit ID). The URL, combined with respective IDs, will be
registered in the dataset as the source of the DICOM tarball, and used
for retrieval by dataset clones.

.. code-block:: bash

>    singularity run -B $STORE_DIR icf.sif deposit_visit_dataset \
>      --store-dir $STORE_DIR --store-url <ICF STORE URL>
> 

Reviewer: ```suggestion
   singularity run -B $STORE_DIR icf.sif deposit_visit_dataset \
     --id <Study ID> <Visit ID> dl-Z03 P000624 --store-dir $STORE_DIR --store-url <ICF STORE URL>

```

Author: Will do that, but without mixing placeholders and values :wink:

This will produce two files, ``<visit ID>_XDLA--refs`` and ``<visit
ID>_XDLA--repo-export`` (text file and zip archive
respectively). Together, they are a representation of a (lightweight)
DataLad dataset, and contain the information necessary to retrieve the
data content with DataLad (but do not contain the data content
itself).

Create a catalog view (optional)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

A catalog page (html+JS rendering of dataset contents generated with
`DataLad catalog`_) can be created for the visit dataset. This is
mostly useful when providing (internal) https access to the datasets.

The following command will create the catalog (or update its content)
and place it in the ``catalog`` folder in the study directory.

.. _DataLad catalog: https://docs.datalad.org/projects/catalog

.. code-block:: bash

   singularity run -B $STORE_DIR icf.sif catalogify_studyvisit_from_meta \
     --store-dir $STORE_DIR --id <project-ID> <visit ID>

> 

Reviewer: I think it would be nice to mention that this catalog needs to be subsequently served, or at least point to the README for further instructions - I naively expected the index.html page to display something and initially thought something was wrong.
